File: i18n.py
"""
Internationalization (i18n) support for SonarQube Checker.

This module provides translations for report generation in multiple languages.
Translations are loaded from translations.yaml file.
"""

import os
import yaml

# Cache for loaded translations
_TRANSLATIONS_CACHE = None


def _load_translations() -> dict:
    """
    Load translations from YAML file.

    Returns:
        Dictionary with translations for all supported languages
    """
    global _TRANSLATIONS_CACHE

    if _TRANSLATIONS_CACHE is not None:
        return _TRANSLATIONS_CACHE

    # Get the directory where this script is located
    script_dir = os.path.dirname(os.path.abspath(__file__))
    translations_file = os.path.join(script_dir, 'translations.yaml')

    try:
        with open(translations_file, 'r', encoding='utf-8') as f:
            _TRANSLATIONS_CACHE = yaml.safe_load(f)
            return _TRANSLATIONS_CACHE
    except FileNotFoundError:
        # Fallback to hardcoded English translations if file not found
        return {
            'en': {
                'report_title': 'SonarQube Analysis Report',
                'generated': 'Generated',
                'project': 'Project',
                'last_analysis': 'Last Analysis',
                'latest_issues': 'Latest Issues',
                'no_analysis_available': 'No analysis available',
                'no_open_issues': 'No open issues found.',
                'severity': 'Severity',
                'message': 'Message',
                'component': 'Component',
                'line': 'Line',
            }
        }


# Load translations on module import
TRANSLATIONS = _load_translations()


def get_translation(key: str, language: str = 'en') -> str:
    """
    Get translated string for the given key and language.

    Args:
        key: Translation key
        language: Language code ('en' or 'ru')

    Returns:
        Translated string, or English fallback if not found
    """
    # Validate language
    if language not in TRANSLATIONS:
        language = 'en'

    # Get translation or fallback to English
    return TRANSLATIONS.get(language, {}).get(key, TRANSLATIONS['en'].get(key, key))


def get_all_translations(language: str = 'en') -> dict:
    """
    Get all translations for a specific language.

    Args:
        language: Language code ('en' or 'ru')

    Returns:
        Dictionary of all translations for the language
    """
    if language not in TRANSLATIONS:
        language = 'en'

    return TRANSLATIONS[language]

File: src/i18n.rs
use clap::ValueEnum;
use once_cell::sync::Lazy;
use serde::Deserialize;
use std::collections::HashMap;

#[derive(Debug, Deserialize)]
pub struct Translations {
    pub en: HashMap<String, String>,
    pub ru: HashMap<String, String>,
}

pub static TRANSLATIONS: Lazy<Translations> = Lazy::new(|| {
    let yaml_content = include_str!("../translations.yaml");
    serde_yaml::from_str(yaml_content).unwrap_or_else(|_| {
        let mut en = HashMap::new();
        en.insert("report_title".to_string(), "SonarQube Analysis Report".to_string());
        en.insert("generated".to_string(), "Generated".to_string());
        en.insert("project".to_string(), "Project".to_string());
        en.insert("last_analysis".to_string(), "Last Analysis".to_string());
        en.insert("latest_issues".to_string(), "Latest Issues".to_string());
        en.insert("no_analysis_available".to_string(), "No analysis available".to_string());
        en.insert("no_open_issues".to_string(), "No open issues found.".to_string());
        en.insert("severity".to_string(), "Severity".to_string());
        en.insert("message".to_string(), "Message".to_string());
        en.insert("component".to_string(), "Component".to_string());
        en.insert("line".to_string(), "Line".to_string());

        let ru = HashMap::new();
        
        Translations { en, ru }
    })
});

#[derive(Debug, Clone, Copy, ValueEnum, PartialEq)]
pub enum Language {
    En,
    Ru,
}

impl Language {
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "ru" => Language::Ru,
            _ => Language::En,
        }
    }
}

pub fn get_translation(key: &str, language: &Language) -> &str {
    let translations = match language {
        Language::En => &TRANSLATIONS.en,
        Language::Ru => &TRANSLATIONS.ru,
    };
    
    translations.get(key)
        .map(|s| s.as_str())
        .or_else(|| TRANSLATIONS.en.get(key).map(|s| s.as_str()))
        .unwrap_or(key)
}